pkg/gardenlet/controller/managedseed/add.go

```diff
@@ -0,0 +1,327 @@
+
+// filterManagedSeed is filtering func for ManagedSeeds that checks if the ManagedSeed references a Shoot scheduled on a Seed,
+// for which the gardenlet is responsible.
+func (p *managedSeedFilterPredicate) filterManagedSeed(obj client.Object) bool {
+	managedSeed, ok := obj.(*seedmanagementv1alpha1.ManagedSeed)
+	if !ok {
+		return false
+	}
+
+	return filterManagedSeedHelper(p.ctx, p.reader, managedSeed, p.gardenNamespace, p.seedName)
+}
+
+// SeedOfManagedSeedFilterPredicate returns the predicate for Seed events.
+func (r *Reconciler) SeedOfManagedSeedFilterPredicate(seedName string) predicate.Predicate {
+	return &seedOfManagedSeedFilterPredicate{
+		reader:          r.GardenClient,
+		gardenNamespace: r.GardenNamespaceGarden,
+		seedName:        seedName,
+	}
+}
+
+type seedOfManagedSeedFilterPredicate struct {
+	ctx             context.Context
+	reader          client.Reader
+	gardenNamespace string
+	seedName        string
+}
+
+func (p *seedOfManagedSeedFilterPredicate) InjectStopChannel(stopChan <-chan struct{}) error {
+	p.ctx = contextutil.FromStopChannel(stopChan)
+	return nil
+}
+
+func (p *seedOfManagedSeedFilterPredicate) Create(e event.CreateEvent) bool {
+	return p.filterSeedOfManagedSeed(e.Object)
+}
+
+func (p *seedOfManagedSeedFilterPredicate) Update(e event.UpdateEvent) bool {
+	return p.filterSeedOfManagedSeed(e.ObjectNew)
+}
+
+func (p *seedOfManagedSeedFilterPredicate) Delete(e event.DeleteEvent) bool {
+	return p.filterSeedOfManagedSeed(e.Object)
+}
+
+func (p *seedOfManagedSeedFilterPredicate) Generic(_ event.GenericEvent) bool { return false }
+
+// filterSeedOfManagedSeed is filtering func for Seeds that checks if the Seed is owned by a ManagedSeed that references a Shoot
+// scheduled on a Seed, for which the gardenlet is responsible.
+func (p *seedOfManagedSeedFilterPredicate) filterSeedOfManagedSeed(obj client.Object) bool {
+	seed, ok := obj.(*gardencorev1beta1.Seed)
+	if !ok {
+		return false
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := p.reader.Get(p.ctx, kutil.Key(p.gardenNamespace, seed.Name), managedSeed); err != nil {
+		return false
+	}
+
+	return filterManagedSeedHelper(p.ctx, p.reader, managedSeed, p.gardenNamespace, p.seedName)
+}
+
+func filterManagedSeedHelper(ctx context.Context, reader client.Reader, managedSeed *seedmanagementv1alpha1.ManagedSeed, gardenNamespace, seedName string) bool {
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		return false
+	}
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := reader.Get(ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		return false
+	}
+
+	if shoot.Spec.SeedName == nil {
+		return false
+	}
+
+	if shoot.Status.SeedName == nil || *shoot.Spec.SeedName == *shoot.Status.SeedName {
+		return *shoot.Spec.SeedName == seedName
+	}
+
+	return *shoot.Status.SeedName == seedName
```

Comment:
Can you do something like in https://github.com/gardener/gardener/pull/7206/files#diff-eb33ba906d976101dfe4f69a634d451fd48618d3bfdf57fe7401aca9e03b9bd8R491 here?